### ISSUE TITLE
Extensible HTTP::Server::Context for libraries

### ISSUE THREAD

**Issue by andrewhamon** [crystal-lang/crystal#3863]
The current state of `HTTP:Server::Context` isn't the greatest when it comes to libraries/frameworks/middlewares that need to add additional context (prototypical example would be adding a `params` method).

Currently it seems like the most common route is to simply monkey patch `HTTP::Server::Context` and add whatever is needed, but this can lead to namespace collisions very easily.

It would be great if there was a way to safely register extensions with HTTP context. There was some brief discussion in gitter: https://gitter.im/crystal-lang/crystal?at=5873e020cbcb28177049b76a

The purpose of this issue is to throw around some ideas to see what we could cook up and determine if its worthwhile. I believe its worth noting that Rack has done just fine so far using a shared namespace... but I hope we can do better :)

**Comment by maiha** [comment]
This is a nice issue. I think this problem is not limited in `HTTP` but in **language specification**.

In **Ruby**, that problem has not been solved for a long time since it was born. Unfortunely, we have same problem in **Crystal** too. Updating **gems** or **shards** easily breaks our apps in Ruby and Crystal. I feel this feature prevents many users from using in production.

It's time to change the negative legacy. For example, **Scala** solved this by using `override` modifier. This feature has helped me many times. I hope that such collision safety will be implemented in **Crystal**. :smile:

**Comment by andrewhamon** [comment]
An `override` modifier would definitely be a welcome addition, but I don't think its enough on its own to solve this issue.

**Comment by asterite** [comment]
> I believe its worth noting that Rack has done just fine so far using a shared namespace

So why do you need to "improve" this?

The same is true for Ruby namespaces: if I declare a module `Foo`, then no one else can do it. But Ruby's ecosystem just does fine with that. Plus, in Crystal you'll immediately get a compile error in the case where I use the same property name as yours.

I personally don't think there's anything we can improve here without drastically changing the language.

**Comment by ysbaddaden** [comment]
Yes, simply having `Context#kemal` or `Context#myapp` accessors for framework specifics seems enough to me. Furthermore, any conflict should quickly raise compilation errors

**Comment by andrewhamon** [comment]
Well, technically in Rack people add to the request hash, not monkey patch. But it's still a shared namespace with the same risks.

Some people in the gitter were suggesting that a hash keyed by classes would be cool, so like `context.extensions[MyLib::Params] # instance of MyLib::Params`

**Comment by asterite** [comment]
But what if I choose `MyLib::Params` for my library too?

There's really no difference between the chance of a clash between:

```cr
context.extensions[MyLib::Params]
```

and

```cr
context.extensions_my_lib_params
```

But there is a difference: the later is probably implemented by adding a class variable to `HTTP::Context`, while the former will go through a hash lookup, which is slightly slower. So using a Hash with a type as a key (which, by the way, can't be expressed right now in the language) has the same chances of a clash, is more verbose to write and is slower.

**Comment by andrewhamon** [comment]
Yes, but the point of doing that would be to encourage namespacong.

I'm not worried that there will be two `context.kemal_params`. I'm worried that because Kemal is using `context.params` with no namespacing of any sort _right now_.

**Comment by asterite** [comment]
But that's OK, `params` represents the URL params. There won't be an extension for Kemal that will also use `params` as a name because that's already covered. There's also a [session](https://github.com/kemalcr/kemal-session/blob/master/src/kemal-session/ext/context.cr#L9-L12) property which provides access to the session values, and nobody else will want to use `session` unless it's a different implementation of a session, in which case you won't use `kemal-session` (so you'll chose one of those).

In fact, if I remember correctly, the whole discussion appeared because of [crux](https://github.com/andrewhamon/crux), which is a router. But since it's a router you won't be using kemal's router (which also handles routing), so it's fine if you name your property `params` because there won't be a clash there. Combining crux with kemal doesn't make sense.

**Comment by RX14** [comment]
This problem would come up if you used two routers which modified `HTTP::Server::Context`. For example if you decided to switch out your routing library to another incrementally you would have two routers in the app for a while, and hit this problem.

**Comment by andrewhamon** [comment]
@asterite a really common example of two frameworks being used at the same time in Ruby is when people mount the admin panel of Resque (which is a Sinatra app) in to their Rails app.

Also, a very important distinction between the Rack hash and Crystal context is that even if two Ruby libs used the exact same name, "params", that field would most likely just be overwritten at runtime once the context passes from one app to another. That can't happen in Crystal.

**Comment by andrewhamon** [comment]
Btw, I'm not a huge fan of the idea of changing the language to support this either :) but I do think that there is room for improvement, either by establishing some conventions around this and maybe by proving some generic context helpers in the standard lib. Defining a Session interface and maybe adding a JSON::Any hash for additional params would go a long way. Rack has a fair number of standard helpers and whatnot which pretty much everyone uses.

**Comment by luislavena** [comment]
I've encountered the issue described but however decided is a co-sharing and design issue and not specifically something that Crystal has to solve.

In my case, I've decided against exposing these router library specific details in `Context` and instead provide my own extension to it using my namespace. For example Beryl exposes `beryl` which allow you to access `params` within it (if you're interfacing with Context directly).

I've also wrapped the original HTTP Context and use `delegate` to give direct access to `params` without having to use `beryl.params` always.

But that way, I keep things isolated from the Context, allowing me to run by router and other router library that might be extending Context without being affected.

Cheers.

**Comment by rishavs** [comment]
reviving this discussion.

My use case is that I need to pass session related data between middlewares and the best way (and how I believe many other modern frameworks do it today) is to be add a store object right in the http context.

While there are workarounds and I am not blocked, the resultant code feel very inelegant where one middleware has to create a side effect object and another has to take this object as an argument.

**Comment by straight-shoota** [comment]
The use case is similar to what [`http-session`](https://github.com/straight-shoota/http-session) deals with and initially used a similar approach. But I was ultimately convinced that it's a better idea to *not* store the session information on the context. Instead, it uses a table to look up the session by ID. This is of course a little bit less performant because the lookup needs to happen twice per request. In relation to what else happens on a request, this is probably completely negligible, though. From an architectural point of view this is an elegant solution because it doesn't contaminate any shared context.

**Comment by luislavena** [comment]
Thanks for sharing that library @straight-shoota, I think is interesting to see different approaches on how that is being handled.

The example you shared makes sense for session loading (creating a cookie, loading from it), but I'm wondering on the overhead of having to query every time for the session information in order to obtain it.

Wouldn't be more efficient a middleware in the server chain that obtains it and exposes it to us without having us to interface with it directly?

Thank you.

**Comment by straight-shoota** [comment]
If you have a local collection of the currently/recently active sessions (for example a local cache), "query every time" is essentially just a relatively simple hash lookup.

**Comment by Sija** [comment]
@straight-shoota Yeah, but that's rarely the case.
